lib/surface/directive/values.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Surface.Directive.Values do
+  use Surface.Directive
+
+  def extract({":values", {:attribute_expr, value, expr_meta}, attr_meta}, meta) do
+    expr_meta = Helpers.to_meta(expr_meta, meta)
+    attr_meta = Helpers.to_meta(attr_meta, meta)
+
+    %AST.Directive{
+      module: __MODULE__,
+      name: :values,
+      value: directive_value(value, expr_meta),
+      meta: attr_meta
+    }
+  end
+
+  def extract(_, _), do: []
+
+  def process(
+        %AST.Directive{
+          value: %AST.AttributeExpr{value: value} = expr,
+          meta: meta
+        },
+        %type{attributes: attributes} = node
+      )
+      when type in [AST.Tag, AST.VoidTag] do
+    attr_names = for %AST.Attribute{name: name} <- attributes, do: name
+
+    new_expr =
+      quote generated: true do
+        for {name, value} <- unquote(value) || [],
+            attr_name = :"phx-value-#{name}",
+            not Enum.member?(unquote(Macro.escape(attr_names)), attr_name) do
+          unless String.Chars.impl_for(value) do
+            message = """
+            invalid value for key "#{inspect(name)}" in attribute ":values".
+
+            Expected a type that implements the String.Chars protocol (e.g. string, boolean, integer, atom, ...), \
+            got: #{inspect(unquote(value))}\
+            """
+
+            IOHelper.runtime_error(message)
```

Comment:
Maybe you could move this code to a separate function, similar to what we do here: https://github.com/surface-ui/surface/blob/71048b56bc391e1e115219cafea5b2934c521d76/lib/surface/directive/events.ex#L53

This way we avoid generating too much code inside `quote`, which makes it easier to debug when things go wrong.

You could call it `validate_values!`.

Reply:
Done!

I must admit that I did try and fail to extract this to a separate function before your hint... quote/unquote is still mindbending to me...